File: ts/project/.sdk/src/cmp/zig/fragment/Main.fragment.zig
// ProjectNameSDK client (generated — mirrors the go/rust Main fragment).

const std = @import("std");
const vs = @import("voxgig-struct");
const h = @import("helpers.zig");
const errmod = @import("error.zig");
const types = @import("types.zig");
const ctxmod = @import("context.zig");
const utility_mod = @import("utility.zig");
const spec_mod = @import("spec.zig");
const config = @import("config.zig");

const Value = h.Value;
const E = errmod.E;
const Context = ctxmod.Context;
const CtxSpec = ctxmod.CtxSpec;
const Utility = utility_mod.Utility;
const Feature = types.Feature;
const OpResult = types.OpResult;
const Spec = spec_mod.Spec;

pub const ProjectNameSDK = struct {
    mode: []const u8 = "live",
    options: Value = .{ .null = {} },
    // NOT named `utility`, and NOT a name `zigVarName` can produce.
    //
    // Entity accessors are generated as methods on this same struct
    // (MainEntity_zig), so an API with an entity called `utility` collided
    // twice: Zig rejects a local binding that shadows the declaration, and
    // `sdk.utility` resolves to the FIELD, which made the generated accessor
    // unreachable.
    //
    // `zigVarName` lowercases every name it is given, so a field spelling with
    // an uppercase letter is one no entity name can ever reach - which is why
    // this is `sdkUtility` and not, say, `util_rt`, a name an entity called
    // `util-rt` would map straight onto. The public reader is `get_utility()`,
    // so this field is internal and free to be spelled this way.
    sdkUtility: *Utility,
    features: std.ArrayList(Feature),
    rootctx: ?*Context = null,

    pub fn new(options: Value) *ProjectNameSDK {
        const sdk = h.A().create(ProjectNameSDK) catch unreachable;
        sdk.* = .{
            .mode = "live",
            .options = h.vnull(),
            .sdkUtility = Utility.new(),
            .features = std.ArrayList(Feature).init(h.A()),
            .rootctx = null,
        };

        // The process-wide config (sdkgen rung L2): read-only on the request
        // path, so every client shares one rather than rebuilding it.
        const cfg = config.shared_config();

        const rootctx = sdk.sdkUtility.make_context(CtxSpec{
            .client = sdk,
            .utility = sdk.sdkUtility,
            .config = cfg,
            .options = options,
            .shared = h.omap(),
        }, null);

        const opts = sdk.sdkUtility.make_options(rootctx);
        sdk.options = opts;

        if (h.veq(h.getpath(&.{ "feature", "test", "active" }, opts), h.vbool(true))) {
            sdk.mode = "test";
        }

        rootctx.options = opts;
        sdk.rootctx = rootctx;

        // Add features in the resolved order (make_options puts an explicit
        // list order first, else defaults to test-first). Ordering matters:
        // the `test` feature installs the base mock transport and the
        // transport features (retry/cache/netsim/proxy/ratelimit) wrap
        // whatever is current, so `test` must be added before them to sit at
        // the base of the transport wrapper chain.
        const feature_opts = h.to_map(h.getp(opts, "feature"));
        const feature_order = h.getpath(&.{ "__derived__", "featureorder" }, opts);
        if (feature_opts == .object and feature_order == .array) {
            for (feature_order.array.data.items) |fname_v| {
                if (fname_v != .string) continue;
                const fname = fname_v.string;
                const fopts = h.getp(feature_opts, fname);
                if (fopts == .object) {
                    if (h.get_bool(fopts, "active") orelse false) {
                        sdk.sdkUtility.feature_add(rootctx, config.make_feature(fname));
                    }
                }
            }
        }

        // Initialize features.
        var snap = std.ArrayList(Feature).init(h.A());
        for (sdk.features.items) |f| snap.append(f) catch {};
        for (snap.items) |f| sdk.sdkUtility.feature_init(rootctx, f);

        sdk.sdkUtility.feature_hook(rootctx, "PostConstruct");

        return sdk;
    }

    pub fn options_map(self: *ProjectNameSDK) Value {
        return h.clone(self.options);
    }

    pub fn get_utility(self: *ProjectNameSDK) *Utility {
        return Utility.copy(self.sdkUtility);
    }

    pub fn get_root_ctx(self: *ProjectNameSDK) *Context {
        return self.rootctx orelse unreachable;
    }

    pub fn prepare(self: *ProjectNameSDK, fetchargs_in: Value) E!Value {

        const fetchargs: Value = switch (fetchargs_in) {
            .object => fetchargs_in,
            else => h.omap(),
        };

        const ctrl: Value = switch (h.to_map(h.getp(fetchargs, "ctrl"))) {
            .object => h.to_map(h.getp(fetchargs, "ctrl")),
            else => h.omap(),
        };

        const ctx = self.sdkUtility.make_context(CtxSpec{
            .opname = "prepare",
            .ctrl = ctrl,
        }, self.get_root_ctx());

        const options = self.options;

        const path = h.get_str(fetchargs, "path") orelse "";
        const method: []const u8 = blk: {
            const m = h.get_str(fetchargs, "method");
            break :blk if (m) |mm| (if (mm.len == 0) "GET" else mm) else "GET";
        };

        const params: Value = switch (h.to_map(h.getp(fetchargs, "params"))) {
            .object => h.to_map(h.getp(fetchargs, "params")),
            else => h.omap(),
        };
        const query: Value = switch (h.to_map(h.getp(fetchargs, "query"))) {
            .object => h.to_map(h.getp(fetchargs, "query")),
            else => h.omap(),
        };

        const headers = self.sdkUtility.prepare_headers(ctx);

        const specmap = h.jo(&.{
            .{ "base", h.getp(options, "base") },
            .{ "prefix", h.getp(options, "prefix") },
            .{ "suffix", h.getp(options, "suffix") },
            .{ "path", h.vstr(path) },
            .{ "method", h.vstr(method) },
            .{ "params", params },
            .{ "query", query },
            .{ "headers", headers },
            .{ "body", h.getp(fetchargs, "body") },
            .{ "step", h.vstr("start") },
        });
        const spec = Spec.make(specmap);
        ctx.spec = spec;

        // Merge user-provided headers.
        if (h.getp(fetchargs, "headers") == .object) {
            const uh = h.getp(fetchargs, "headers");
            var it = uh.object.iterator();
            while (it.next()) |kv| h.setp(spec.headers, kv.key_ptr.*, kv.value_ptr.*);
        }

        _ = try self.sdkUtility.prepare_auth(ctx);

        return self.sdkUtility.make_fetch_def(ctx);
    }

    // Raw endpoint access is operator-controllable, like every entity op.
    // Blocking it means denying BOTH the 'direct' and 'graphql' tokens,
    // since either one reaches the same endpoint.
    pub fn direct(self: *ProjectNameSDK, fetchargs_in: Value) Value {
        if (!self.op_allowed("direct")) return self.op_denied("direct");

        return self.raw_request(fetchargs_in);
    }

    // Is this raw-access op permitted by the SDK's allow.op option?
    fn op_allowed(self: *ProjectNameSDK, op: []const u8) bool {
        const allow: []const u8 = switch (h.getpath(&.{ "allow", "op" }, self.options)) {
            .string => |s| s,
            else => "",
        };
        return std.mem.indexOf(u8, allow, op) != null;
    }

    fn op_denied(self: *ProjectNameSDK, op: []const u8) Value {
        const allow: []const u8 = switch (h.getpath(&.{ "allow", "op" }, self.options)) {
            .string => |s| s,
            else => "",
        };
        const msg = std.fmt.allocPrint(h.A(),
            "ProjectNameSDK: {s}: operation not allowed by" ++
            " SDK option allow.op value: \"{s}\"", .{ op, allow }) catch "";
        return h.jo(&.{
            .{ "ok", h.vbool(false) },
            .{ "err", h.vstr(msg) },
        });
    }

    // Ungated request path shared by direct and graphql, each of which checks
    // its own allow.op token first. Private, rather than a flag on fetchargs:
    // a caller-supplied marker would let anyone opt straight back out of the
    // gate by passing it.
    fn raw_request(self: *ProjectNameSDK, fetchargs_in: Value) Value {

        const fetchdef = self.prepare(fetchargs_in) catch {
            return h.jo(&.{
                .{ "ok", h.vbool(false) },
                .{ "err", h.vstr(if (self.rootctx.?.pending_err) |e| e.msg else "prepare failed") },
            });
        };

        const fetchargs: Value = switch (fetchargs_in) {
            .object => fetchargs_in,
            else => h.omap(),
        };
        const ctrl: Value = switch (h.to_map(h.getp(fetchargs, "ctrl"))) {
            .object => h.to_map(h.getp(fetchargs, "ctrl")),
            else => h.omap(),
        };

        const ctx = self.sdkUtility.make_context(CtxSpec{
            .opname = "direct",
            .ctrl = ctrl,
        }, self.get_root_ctx());

        const url = h.get_str(fetchdef, "url") orelse "";
        const fetched = self.sdkUtility.fetch(ctx, url, fetchdef) catch {
            return h.jo(&.{
                .{ "ok", h.vbool(false) },
                .{ "err", h.vstr(if (ctx.pending_err) |e| e.msg else "fetch failed") },
            });
        };

        if (h.is_noval(fetched)) {
            return h.jo(&.{
                .{ "ok", h.vbool(false) },
                .{ "err", h.vstr("response: undefined") },
            });
        }

        if (fetched == .object) {
            const status = h.to_int(h.getp(fetched, "status"));
            const headers = h.getp(fetched, "headers");

            const content_length: []const u8 = switch (h.getp(headers, "content-length")) {
                .string => |s| s,
                .integer => |n| std.fmt.allocPrint(h.A(), "{d}", .{n}) catch "",
                else => "",
            };
            const no_body = status == 204 or status == 304 or std.mem.eql(u8, content_length, "0");

            const json_data: Value = if (no_body) h.vnull() else blk: {
                const jf = h.getp(fetched, "json");
                break :blk if (jf == .function) h.call_json(jf) else h.vnull();
            };

            return h.jo(&.{
                .{ "ok", h.vbool(200 <= status and status < 300) },
                .{ "status", h.vnum(status) },
                .{ "headers", headers },
                .{ "data", json_data },
            });
        }

        return h.jo(&.{
            .{ "ok", h.vbool(false) },
            .{ "err", h.vstr("invalid response type") },
        });
    }

    // Raw GraphQL access: the pressure valve that makes the generated
    // surface's deliberate omissions (per-call selection sets, typed filter
    // builders, batching, subscriptions) livable — the whole schema stays
    // reachable.
    //
    // Thin wrapper over the same prepare/fetch path direct uses, with the one
    // thing raw direct cannot do for GraphQL: a GraphQL failure rides HTTP
    // 200 as a top-level `errors` array, so status alone would report a
    // failed query as ok.
    //
    // NOTE: like direct, this bypasses the feature pipeline — no retry,
    // ratelimit or paging features apply.
    pub fn graphql(
        self: *ProjectNameSDK, query: []const u8, variables: Value, ctrl: Value,
    ) Value {
        if (!self.op_allowed("graphql")) return self.op_denied("graphql");

        const vars: Value = switch (variables) {
            .object => variables,
            else => h.omap(),
        };
        const ctl: Value = switch (ctrl) {
            .object => ctrl,
            else => h.omap(),
        };

        const res = self.raw_request(h.jo(&.{
            .{ "method", h.vstr("POST") },
            .{ "headers", h.jo(&.{.{ "content-type", h.vstr(utility_mod.GRAPHQL_CONTENT_TYPE) }}) },
            .{ "body", h.jo(&.{ .{ "query", h.vstr(query) }, .{ "variables", vars } }) },
            .{ "ctrl", ctl },
        }));

        if (res != .object) return res;

        // Errors are read BEFORE any status check: a GraphQL parse or
        // validation failure comes back as HTTP 400 carrying the standard
        // { errors: [...] } body, and the raw path represents a non-2xx as
        // ok:false with no err — so returning early on status would discard
        // the server's own diagnostics, which are the only useful part of
        // that response.
        const errors = h.getp(h.getp(res, "data"), "errors");

        if (errors == .array and 0 < errors.array.data.items.len) {
            const first = errors.array.data.items[0];
            const m: []const u8 = switch (h.getp(first, "message")) {
                .string => |x| if (x.len == 0) "graphql error" else x,
                else => "graphql error",
            };
            const msg = std.fmt.allocPrint(h.A(),
                "ProjectNameSDK: graphql: {s}", .{m}) catch "";
            h.setp(res, "ok", h.vbool(false));
            h.setp(res, "err", h.vstr(msg));
            h.setp(res, "graphql", errors);
        }

        return res;
    }

    // <[SLOT]>
};

pub fn test_sdk(testopts_in: Value, sdkopts_in: Value) *ProjectNameSDK {
    const sdkopts: Value = switch (sdkopts_in) {
        .object => h.clone(sdkopts_in),
        else => h.omap(),
    };

    const testopts: Value = switch (testopts_in) {
        .object => h.clone(testopts_in),
        else => h.omap(),
    };
    h.setp(testopts, "active", h.vbool(true));

    // set_path mutates `sdkopts` in place; keep the ROOT (gotcha #8 — do not
    // rebind to the return of setpath).
    h.setpath(sdkopts, &.{ "feature", "test" }, testopts);

    const sdk = ProjectNameSDK.new(sdkopts);
    sdk.mode = "test";

    return sdk;
}

File: ts/project/.sdk/src/cmp/c/fragment/Main.fragment.c
// ProjectNameSDK client (generated — mirrors the rust Main fragment).

#include "api.h"

#include <stdio.h>   // snprintf
#include <stdlib.h>
#include <string.h>

ProjectNameSDK* projectname_sdk_new(voxgig_value* options) {
  ProjectNameSDK* sdk = (ProjectNameSDK*)calloc(1, sizeof(ProjectNameSDK));
  sdk->mode = strdup("live");
  sdk->options = voxgig_new_undef();
  sdk->utility = utility_new();
  sdk->features = NULL;
  sdk->features_len = 0;
  sdk->features_cap = 0;
  sdk->rootctx = NULL;

  voxgig_value* config = make_config();

  CtxSpec cs;
  memset(&cs, 0, sizeof(cs));
  cs.client = sdk;
  cs.utility = sdk->utility;
  cs.config = config;
  cs.options = options ? options : voxgig_new_undef();
  cs.shared = voxgig_new_map();
  Context* rootctx = make_context_util(cs, NULL);

  voxgig_value* opts = make_options_util(rootctx);
  sdk->options = v_share(opts);

  voxgig_value* testactive;
  {
    const char* keys[4] = {"feature", "test", "active", NULL};
    testactive = getpath_c(opts, keys);
  }
  if (voxgig_is_bool(testactive) && voxgig_as_bool(testactive)) {
    free(sdk->mode);
    sdk->mode = strdup("test");
  }

  rootctx->options = v_share(opts);
  sdk->rootctx = rootctx;

  // Add features in the resolved order (make_options puts an explicit list
  // order first, else defaults to test-first). Ordering matters: the `test`
  // feature installs the base mock transport and the transport features
  // (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
  // must be added before them to sit at the base of the transport chain.
  voxgig_value* feature_opts = to_map(getp(opts, "feature"));
  voxgig_value* feature_order = getpath2(opts, "__derived__", "featureorder");
  if (voxgig_is_map(feature_opts) && v_is_list(feature_order)) {
    voxgig_list* order = voxgig_as_list(feature_order);
    for (size_t i = 0; i < order->len; i++) {
      voxgig_value* fname_v = order->items[i];
      if (!v_is_str(fname_v)) continue;
      const char* fname = voxgig_as_string(fname_v);
      voxgig_value* fopts = getp(feature_opts, fname);
      if (voxgig_is_map(fopts)) {
        bool active = false;
        if (get_bool(fopts, "active", &active) && active) {
          feature_add_util(rootctx, make_feature(fname));
        }
      }
    }
  }

  // Initialize features.
  size_t n = sdk->features_len;
  for (size_t i = 0; i < n; i++) {
    feature_init_util(rootctx, sdk->features[i]);
  }

  feature_hook_util(rootctx, "PostConstruct");

  return sdk;
}

voxgig_value* sdk_prepare(ProjectNameSDK* sdk, voxgig_value* fetchargs, PNError** err) {
  *err = NULL;
  Utility* utility = sdk->utility;
  (void)utility;

  fetchargs = voxgig_is_map(fetchargs) ? fetchargs : voxgig_new_map();

  voxgig_value* ctrl = to_map(getp(fetchargs, "ctrl"));
  if (!voxgig_is_map(ctrl)) ctrl = voxgig_new_map();

  CtxSpec cs;
  memset(&cs, 0, sizeof(cs));
  cs.opname = "prepare";
  cs.ctrl = ctrl;
  Context* ctx = make_context_util(cs, sdk_get_root_ctx(sdk));

  voxgig_value* options = v_share(sdk->options);

  const char* path = get_str(fetchargs, "path");
  path = path ? path : "";
  const char* method = get_str(fetchargs, "method");
  if (!method || method[0] == '\0') method = "GET";

  voxgig_value* params = to_map(getp(fetchargs, "params"));
  if (!voxgig_is_map(params)) params = voxgig_new_map();
  voxgig_value* query = to_map(getp(fetchargs, "query"));
  if (!voxgig_is_map(query)) query = voxgig_new_map();

  voxgig_value* headers = prepare_headers_util(ctx);

  voxgig_value* specmap = cmap(10,
    "base", getp(options, "base"),
    "prefix", getp(options, "prefix"),
    "suffix", getp(options, "suffix"),
    "path", v_str(path),
    "method", v_str(method),
    "params", params,
    "query", query,
    "headers", headers,
    "body", getp(fetchargs, "body"),
    "step", v_str("start"));
  Spec* spec = spec_new(specmap);
  ctx->spec = spec;

  // Merge user-provided headers.
  voxgig_value* uh = getp(fetchargs, "headers");
  if (voxgig_is_map(uh)) {
    voxgig_map* m = voxgig_as_map(uh);
    for (size_t i = 0; i < m->len; i++) {
      setp(spec->headers, m->entries[i].key, voxgig_retain(m->entries[i].value));
    }
  }

  prepare_auth_util(ctx, err);
  if (*err) return NULL;

  return make_fetch_def_util(ctx, err);
}

static voxgig_value* err_map(const char* msg) {
  return cmap(2, "ok", v_bool(false), "err", v_str(msg));
}

// Is this raw-access op permitted by the SDK's allow.op option?
static bool sdk_op_allowed(ProjectNameSDK* sdk, const char* op) {
  voxgig_value* allow_op = getpath2(sdk->options, "allow", "op");
  if (!voxgig_is_string(allow_op)) return false;
  return NULL != strstr(voxgig_as_string(allow_op), op);
}

static voxgig_value* sdk_op_denied(ProjectNameSDK* sdk, const char* op) {
  voxgig_value* allow_op = getpath2(sdk->options, "allow", "op");
  const char* allow = voxgig_is_string(allow_op) ? voxgig_as_string(allow_op) : "";
  char msg[512];
  snprintf(msg, sizeof(msg),
    "ProjectNameSDK: %s: operation not allowed by"
    " SDK option allow.op value: \"%s\"", op, allow);
  return err_map(msg);
}

// Ungated request path shared by sdk_direct and sdk_graphql, each of which
// checks its own allow.op token first. Static, rather than a flag on
// fetchargs: a caller-supplied marker would let anyone opt straight back out
// of the gate by passing it.
static voxgig_value* sdk_raw_request(
  ProjectNameSDK* sdk, voxgig_value* fetchargs, PNError** err) {
  *err = NULL;
  Utility* utility = sdk->utility;

  PNError* perr = NULL;
  voxgig_value* fetchdef = sdk_prepare(sdk, fetchargs, &perr);
  if (perr) {
    return err_map(perr->msg);
  }

  voxgig_value* ctrl = to_map(getp(fetchargs, "ctrl"));
  if (!voxgig_is_map(ctrl)) ctrl = voxgig_new_map();

  CtxSpec cs;
  memset(&cs, 0, sizeof(cs));
  cs.opname = "direct";
  cs.ctrl = ctrl;
  Context* ctx = make_context_util(cs, sdk_get_root_ctx(sdk));

  const char* url = get_str(fetchdef, "url");
  url = url ? url : "";
  PNError* ferr = NULL;
  voxgig_value* fetched = utility_fetch(utility, ctx, url, fetchdef, &ferr);
  if (ferr) {
    return err_map(ferr->msg);
  }

  if (v_is_noval(fetched) || v_is_null(fetched)) {
    return err_map("response: undefined");
  }

  if (voxgig_is_map(fetched)) {
    int64_t status = to_int(getp(fetched, "status"));
    voxgig_value* headers = getp(fetched, "headers");

    voxgig_value* cl = getp(headers, "content-length");
    char clbuf[32];
    clbuf[0] = '\0';
    if (voxgig_is_string(cl)) {
      snprintf(clbuf, sizeof(clbuf), "%s", voxgig_as_string(cl));
    } else if (voxgig_is_number(cl)) {
      snprintf(clbuf, sizeof(clbuf), "%lld", (long long)to_int(cl));
    }
    bool no_body = (status == 204 || status == 304 || strcmp(clbuf, "0") == 0);

    voxgig_value* json_data;
    if (no_body) {
      json_data = voxgig_new_undef();
    } else {
      voxgig_value* jf = getp(fetched, "json");
      json_data = voxgig_is_func(jf) ? call_json(jf) : voxgig_new_undef();
    }

    return cmap(4,
      "ok", v_bool(status >= 200 && status < 300),
      "status", v_num((double)status),
      "headers", v_share(headers),
      "data", json_data);
  }

  return err_map("invalid response type");
}

// Raw endpoint access is operator-controllable, like every entity op.
// Blocking it means denying BOTH the 'direct' and 'graphql' tokens, since
// either one reaches the same endpoint.
voxgig_value* sdk_direct(ProjectNameSDK* sdk, voxgig_value* fetchargs, PNError** err) {
  *err = NULL;

  if (!sdk_op_allowed(sdk, "direct")) {
    return sdk_op_denied(sdk, "direct");
  }

  return sdk_raw_request(sdk, fetchargs, err);
}

// Raw GraphQL access: the pressure valve that makes the generated surface's
// deliberate omissions (per-call selection sets, typed filter builders,
// batching, subscriptions) livable — the whole schema stays reachable.
//
// Thin wrapper over the same prepare/fetch path sdk_direct uses, with the one
// thing raw direct cannot do for GraphQL: a GraphQL failure rides HTTP 200 as
// a top-level `errors` array, so status alone would report a failed query as
// ok.
//
// NOTE: like sdk_direct, this bypasses the feature pipeline — no retry,
// ratelimit or paging features apply.
voxgig_value* sdk_graphql(ProjectNameSDK* sdk, const char* query,
                          voxgig_value* variables, voxgig_value* ctrl,
                          PNError** err) {
  *err = NULL;

  if (!sdk_op_allowed(sdk, "graphql")) {
    return sdk_op_denied(sdk, "graphql");
  }

  voxgig_value* vars = voxgig_is_map(variables) ? v_share(variables) : voxgig_new_map();
  voxgig_value* ctl = voxgig_is_map(ctrl) ? v_share(ctrl) : voxgig_new_map();

  voxgig_value* fetchargs = cmap(4,
    "method", v_str("POST"),
    "headers", cmap(1, "content-type", v_str(GRAPHQL_CONTENT_TYPE)),
    "body", cmap(2, "query", v_str(query ? query : ""), "variables", vars),
    "ctrl", ctl);

  voxgig_value* res = sdk_raw_request(sdk, fetchargs, err);
  if (*err || !voxgig_is_map(res)) return res;

  // Errors are read BEFORE any status check: a GraphQL parse or validation
  // failure comes back as HTTP 400 carrying the standard { errors: [...] }
  // body, and the raw path represents a non-2xx as ok:false with no err — so
  // returning early on status would discard the server's own diagnostics,
  // which are the only useful part of that response.
  voxgig_value* errors = getp(getp(res, "data"), "errors");

  if (voxgig_is_list(errors) && 0 < voxgig_as_list(errors)->len) {
    voxgig_value* first = voxgig_as_list(errors)->items[0];
    const char* m = get_str(first, "message");
    if (!m || '\0' == m[0]) m = "graphql error";
    char msg[512];
    snprintf(msg, sizeof(msg), "ProjectNameSDK: graphql: %s", m);
    setp(res, "ok", v_bool(false));
    setp(res, "err", v_str(msg));
    setp(res, "graphql", v_share(errors));
  }

  return res;
}

// <[SLOT]>

ProjectNameSDK* test_sdk(voxgig_value* testopts, voxgig_value* sdkopts) {
  sdkopts = voxgig_is_map(sdkopts) ? voxgig_clone(sdkopts) : voxgig_new_map();
  testopts = voxgig_is_map(testopts) ? voxgig_clone(testopts) : voxgig_new_map();
  setp(testopts, "active", v_bool(true));

  // set_path mutates sdkopts in place; discard the return (keep the ROOT).
  voxgig_value* path = clist(2, v_str("feature"), v_str("test"));
  voxgig_setpath(sdkopts, path, testopts, NULL);

  ProjectNameSDK* sdk = projectname_sdk_new(sdkopts);
  free(sdk->mode);
  sdk->mode = strdup("test");
  return sdk;
}
